pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,263 @@
-	req := e.clientSet.CoreV1().RESTClient().Post().
-		Resource("pods").
-		Name(pod.Name).
-		Namespace(pod.Namespace).
-		SubResource("exec").
-		Param("container", containerName)
-
-	req.VersionedParams(&corev1.PodExecOptions{
-		Container: containerName,
-		Command:   command,
-		Stdin:     false,
-		Stdout:    true,
-		Stderr:    true,
-		TTY:       false,
-	}, scheme.ParameterCodec)
-
-	execSPDY, err := remotecommand.NewSPDYExecutor(e.config, "POST", req.URL())
-	if err != nil {
-		e.log.Error(err, "failed building SPDY (shell) executor")
-		return "", "", err
+	// create a context for the fence agent command that the controller can cancel
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	routine := routine{
+		cancel: cancel,
 	}
+	e.routines[uid] = &routine
 
-	// Execute the Post request for SPDY (shell) executor
-	err = execSPDY.Stream(remotecommand.StreamOptions{
-		Stdout: &stdoutBuf,
-		Stderr: &stderrBuf,
-		Tty:    false,
-	})
-	if err != nil {
-		e.log.Error(err, "Failed to run exec command", "stdout", stdoutBuf.String(), "stderr", stderrBuf.String())
+	go e.fenceAgentRoutine(cancellableCtx, uid, command, retryCount, retryInterval, timeout)
+}
+
+func (e *Executer) fenceAgentRoutine(ctx context.Context, uid types.UID, command []string, retryCount int, retryInterval, timeout time.Duration) {
+	// run the command and update the status
+	retryErr, cmdErr := e.runWithRetry(ctx, uid, command, retryCount, retryInterval, timeout)
+	if retryErr != nil {
+		switch {
+		case errors.Is(retryErr, context.Canceled):
+			e.log.Info("fence agent context canceled. Nothing to do")
+			return
+		case wait.Interrupted(retryErr):
+			e.log.Info("fence agent context timed out")
+		default:
+			e.log.Error(retryErr, "fence agent retry error")
+		}
+	}
+
+	retryErr = e.updateStatusWithRetry(ctx, uid, cmdErr)
+	if retryErr != nil {
+		switch {
+		case wait.Interrupted(retryErr):
+			e.log.Info("status context timed out")
+		default:
+			e.log.Error(retryErr, "status retry error")
+		}
+	}
+}
+
+func (e *Executer) runWithRetry(ctx context.Context, uid types.UID, command []string, retryCount int, retryInterval, timeout time.Duration) (retryErr, faErr error) {
+	// Run the command with an exponantial backoff retry to handle the following cases:
+	// - the command fails: the command is retried until the retryCount is reached
+	// - the command times out: the command is retried until the retryCount is reached
+	// - the FA context times out: the command is cancelled and the status is updated
+	// - the FA context is cancelled: the command is cancelled and the status is not updated
+	// - the command succeeds: the command is not retried and the status is updated
+
+	// Linear backoff
+	backoff := wait.Backoff{
+		Steps:    retryCount,
+		Duration: retryInterval,
+		Factor:   1.0,
+	}
+
+	e.log.Info("fence agent start", "uid", uid, "command", command, "retryCount", retryCount, "retryInterval", retryInterval, "timeout", timeout)
```

Comment:
Don't expose the command in the log, it could include credentials for the API that the fence agent is using. It is not something we want to expose. We could print the fence agent name and the node/cr name for unique and easier-to-identify prints

Reply:
I'll remove the command from the log.

Regarding the suggested change, I prefer to keep it as is. It is a mark in the log to show when the command starts and also it didn't started _yet_, so "is started" is not correct.